### PR TITLE
Re-implement GodotPayment Android plugin using the Google Play Billing library

### PR DESCRIPTION
**Pull request by timoschwarzer**
### What does this PR do?
Re-implement the GodotPayment Android plugin to migrate away from AIDL ("Accessing Google Play Billing using AIDL is deprecated, and all integrations must use the Google Play Billing Library in the future" [source](https://developer.android.com/google/play/billing/migrate)) and to using signals. This also enables the usage of in-app subscriptions.

**This breaks compatibility** but we already broke compatibility by renaming `GodotPayments` to `GodotPayment` for 3.2.2 and the changes should be fairly trivial to implement in existing games (I'll write docs once this PR is done).

### Example code using the new API

Some methods are missing from this example, I'll add them later... (purchasing items, consuming items, querying SKU details)

```gdscript
func _ready():
	if Engine.has_singleton("GodotPayment"):
		payment = Engine.get_singleton("GodotPayment")
		payment.connect('connected', self, '_on_connected') # No params
		payment.connect('disconnected', self, '_on_disconnected') # No params
		payment.connect('connect_error', self, '_on_connect_error') # Response ID (int), Debug message (string)
		payment.connect('purchases_updated', self, '_on_purchases_updated') # Purchases (Dictionary[])
		payment.connect('purchase_error', self, '_on_purchase_error') # Response ID (int), Debug message (string)
		payment.connect('sku_details_query_completed', self, '_on_sku_details_query_completed') # SKUs (Dictionary[])
		payment.connect('sku_details_query_error', self, '_on_sku_details_query_error') # Response ID (int), Debug message (string), Queried SKUs (string[])
		payment.connect('purchase_acknowledged', self, '_on_purchase_acknowledged') # Purchase token (string)
		payment.connect('purchase_acknowledgement_error', self, '_on_purchase_acknowledgement_error') # Response ID (int), Debug message (string), Purchase token (string)
		payment.connect('purchase_consumed', self, '_on_purchase_consumed') # Purchase token (string)
		payment.connect('purchase_consumption_error', self, '_on_purchase_consumption_error') # Response ID (int), Debug message (string), Purchase token (string)
		payment.startConnection()
	else:
		print("GodotPayment singleton is only available on Android devices.")

func _on_connected():
	var query = payment.queryPurchases('subs')
	if query.status == OK:
		for purchase in query.purchases:
			print('The player has purchased: ' + str(purchase.sku))
			if !purchase.is_acknowledged:
				print('Purchase ' + str(purchase.sku) + ' has not been acknowledged...')
				payment.acknowledgePurchase(purchase.purchase_token)

# ...all the other callbacks...
```

### Current problems

~I found no way to conditionally add a Gradle dependency to the custom Android build. `com.android.billingclient:billing:2.2.1` should only be added if the module is enabled in the project export dialog because of 1. size overhead and 2. it automatically adding the `BILLING` permission to the app.
My current workaround is to create a file `<project dir>/android/billing-client/gradle.conf` with the following content to include the library in the final build:~

```
[dependencies]
    implementation 'com.android.billingclient:billing:2.2.1'
```

### Dependencies

If you want to test this, make sure that https://github.com/godotengine/godot/pull/39029 is merged or merge it yourself before as this PR depends on it.

### To-Do

- [x] Find a solution to the Google Play Billing library inclusion problem
- [x] Fix code style to match project code style
- [ ] Write migration guide to make it clear to users that it's now _required_ to acknowledge purchases and subscriptions
- [x] Update example IAP project
- [ ] Update docs

### CODE REVIEW

**Comment by timoschwarzer** [comment]
Maybe there is already a solution to the library inclusion problem in the new Android plugin system that I didn't find... (CC: @m4gr3d)

**Comment by m4gr3d** [comment]
> Maybe there is already a solution to the library inclusion problem in the new Android plugin system that I didn't find... (CC: @m4gr3d)

@timoschwarzer With the new plugin system, you can define a [configuration file](https://docs.godotengine.org/en/stable/tutorials/plugins/android/android_plugin.html#building-a-android-plugin) that can specify the dependencies needed by the plugin.
Since `GodotPayment` is a built-in plugin, its configuration is defined [here](https://github.com/godotengine/godot/blob/7c332d93865d98e8f8f50bc0d5f5a7c2c334f3fc/platform/android/plugin/godot_plugin_config.h#L88), so you just need to update it with the needed dependency.

File: platform/android/java/plugins/godotpayment/build.gradle [hunk 1]
```diff
@@ -21,6 +21,7 @@ android {
 dependencies {
```
Comment (m4gr3d) [inline platform/android/java/plugins/godotpayment/build.gradle:21]: With all the code deletion, seems you should also be able to remove the `useLibrary 'org.apache.http.legacy'`.

**Review by m4gr3d — CHANGES_REQUESTED** [review]
I'm guessing this is not opened against the `master` branch so that you can test and validate?

**Comment by timoschwarzer** [comment]
> With the new plugin system, you can define a [configuration file](https://docs.godotengine.org/en/stable/tutorials/plugins/android/android_plugin.html#building-a-android-plugin) that can specify the dependencies needed by the plugin.

Thank you for the hint @m4gr3d, that works :)! Although I didn't find a better way to create an inline `Vector<String>` than this:

```
/*
 * Set of prebuilt plugins.
 */
static const PluginConfig GODOT_PAYMENT = {
	/*.valid_config =*/true,
	/*.name =*/"GodotPayment",
	/*.binary_type =*/"local",
	/*.binary =*/"res://android/build/libs/plugins/GodotPayment.release.aar",
	/*.local_dependencies =*/{},
	/*.remote_dependencies =*/String("com.android.billingclient:billing:2.2.1").split("|"),
	/*.custom_maven_repos =*/{}
};
```

There has to be a better way...

**Comment by timoschwarzer** [comment]
> I'm guessing this is not opened against the `master` branch so that you can test and validate?

This is opened against 3.2 for now because [@akien-mga said](https://github.com/godotengine/godot/issues/30809#issuecomment-633458313) that we can break compatibility now with 3.2.2 if necessary as we already kind of broke it by renaming GodotPayments to GodotPayment. I can also imagine Google rejecting apps using the old AIDL API in the near future because their warnings are quite bold. ([1](https://developer.android.com/google/play/billing/api), [2](https://developer.android.com/google/play/billing/migrate))

When everything works in 3.2, I'll start testing with master :)

**Comment by timoschwarzer** [comment]
@m4gr3d Thanks a lot for the review! All valid points and done :)

**Review by m4gr3d — APPROVED** [review]
The code looks good to me!

**Comment by akien-mga** [comment]
There are some code styling issues: https://travis-ci.org/github/godotengine/godot/jobs/691014967

You can set up `clang-format` locally to apply changes automatically: https://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html

**Comment by timoschwarzer** [comment]
@akien-mga fixed :+1: 

@m4gr3d are there significant differences between the Android build systems in 3.2 and master? I'd start the PR for 4.0 now :)

**Comment by m4gr3d** [comment]
> @m4gr3d are there significant differences between the Android build systems in 3.2 and master? I'd start the PR for 4.0 now :)

@timoschwarzer  No the changes should be identical.
The previous `v0` system is completely deprecated and removed in `4.0`, but the [`v1` system introduced in this upcoming release](https://docs.godotengine.org/en/stable/tutorials/plugins/android/android_plugin.html) and the one you're using, is made forward-compatible with `4.0`.

**Comment by akien-mga** [comment]
Thanks!

**Comment by volzhs** [comment]
`Engine.has_singleton("GodotPayment")` now returns `false` even running on Android device.
I compiled editor and android template with this PR merged.

**Comment by volzhs** [comment]
I got error log from adb logcat.
```
05-27 13:07:34.469: I/GodotPluginRegistry(22888): Initializing Godot plugin GodotPayment
05-27 13:07:34.470: I/zygote64(22888): Rejecting re-init on previously-failed class java.lang.Class<org.godotengine.godot.plugin.payment.GodotPayment>: java.lang.NoClassDefFoundError: Failed resolution of: Lcom/android/billingclient/api/PurchasesUpdatedListener;
05-27 13:07:34.470: I/zygote64(22888):   at java.lang.Class java.lang.Class.classForName(java.lang.String, boolean, java.lang.ClassLoader) (Class.java:-2)
05-27 13:07:34.470: I/zygote64(22888):   at java.lang.Class java.lang.Class.forName(java.lang.String, boolean, java.lang.ClassLoader) (Class.java:453)
05-27 13:07:34.470: I/zygote64(22888):   at java.lang.Class java.lang.Class.forName(java.lang.String) (Class.java:378)
05-27 13:07:34.470: I/zygote64(22888):   at void org.godotengine.godot.plugin.GodotPluginRegistry.loadPlugins(org.godotengine.godot.Godot) (GodotPluginRegistry.java:173)
05-27 13:07:34.470: I/zygote64(22888):   at void org.godotengine.godot.plugin.GodotPluginRegistry.<init>(org.godotengine.godot.Godot) (GodotPluginRegistry.java:71)
05-27 13:07:34.470: I/zygote64(22888):   at org.godotengine.godot.plugin.GodotPluginRegistry org.godotengine.godot.plugin.GodotPluginRegistry.initializePluginRegistry(org.godotengine.godot.Godot) (GodotPluginRegistry.java:103)
05-27 13:07:34.470: I/zygote64(22888):   at void org.godotengine.godot.Godot.onCreate(android.os.Bundle) (Godot.java:574)
05-27 13:07:34.470: I/zygote64(22888):   at void android.app.Activity.performCreate(android.os.Bundle, android.os.PersistableBundle) (Activity.java:7009)
05-27 13:07:34.470: I/zygote64(22888):   at void android.app.Activity.performCreate(android.os.Bundle) (Activity.java:7000)
05-27 13:07:34.470: I/zygote64(22888):   at void android.app.Instrumentation.callActivityOnCreate(android.app.Activity, android.os.Bundle) (Instrumentation.java:1214)
05-27 13:07:34.470: I/zygote64(22888):   at android.app.Activity android.app.ActivityThread.performLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent) (ActivityThread.java:2731)
05-27 13:07:34.470: I/zygote64(22888):   at void android.app.ActivityThread.handleLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:2856)
05-27 13:07:34.470: I/zygote64(22888):   at void android.app.ActivityThread.-wrap11(android.app.ActivityThread, android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:-1)
05-27 13:07:34.470: I/zygote64(22888):   at void android.app.ActivityThread$H.handleMessage(android.os.Message) (ActivityThread.java:1589)
05-27 13:07:34.470: I/zygote64(22888):   at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:106)
05-27 13:07:34.470: I/zygote64(22888):   at void android.os.Looper.loop() (Looper.java:164)
05-27 13:07:34.470: I/zygote64(22888):   at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:6494)
05-27 13:07:34.470: I/zygote64(22888):   at java.lang.Object java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[]) (Method.java:-2)
05-27 13:07:34.470: I/zygote64(22888):   at void com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run() (RuntimeInit.java:438)
05-27 13:07:34.470: I/zygote64(22888):   at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:807)
05-27 13:07:34.470: I/zygote64(22888): Caused by: java.lang.ClassNotFoundException: Didn't find class "com.android.billingclient.api.PurchasesUpdatedListener" on path: DexPathList[[zip file "/data/app/com.test-s1tfq_9jXDaeuohJv90BvQ==/base.apk"],nativeLibraryDirectories=[/data/app/com.test-s1tfq_9jXDaeuohJv90BvQ==/lib/arm64, /data/app/com.test-s1tfq_9jXDaeuohJv90BvQ==/base.apk!/lib/arm64-v8a, /system/lib64, /vendor/lib64]]
05-27 13:07:34.470: I/zygote64(22888):   at java.lang.Class dalvik.system.BaseDexClassLoader.findClass(java.lang.String) (BaseDexClassLoader.java:125)
05-27 13:07:34.470: I/zygote64(22888):   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String, boolean) (ClassLoader.java:379)
05-27 13:07:34.470: I/zygote64(22888):   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String) (ClassLoader.java:312)
05-27 13:07:34.470: I/zygote64(22888):   at java.lang.Class java.lang.Class.classForName(java.lang.String, boolean, java.lang.ClassLoader) (Class.java:-2)
05-27 13:07:34.470: I/zygote64(22888):   at java.lang.Class java.lang.Class.forName(java.lang.String, boolean, java.lang.ClassLoader) (Class.java:453)
05-27 13:07:34.470: I/zygote64(22888):   at java.lang.Class java.lang.Class.forName(java.lang.String) (Class.java:378)
05-27 13:07:34.470: I/zygote64(22888):   at void org.godotengine.godot.plugin.GodotPluginRegistry.loadPlugins(org.godotengine.godot.Godot) (GodotPluginRegistry.java:173)
05-27 13:07:34.470: I/zygote64(22888):   at void org.godotengine.godot.plugin.GodotPluginRegistry.<init>(org.godotengine.godot.Godot) (GodotPluginRegistry.java:71)
05-27 13:07:34.470: I/zygote64(22888):   at org.godotengine.godot.plugin.GodotPluginRegistry org.godotengine.godot.plugin.GodotPluginRegistry.initializePluginRegistry(org.godotengine.godot.Godot) (GodotPluginRegistry.java:103)
05-27 13:07:34.470: I/zygote64(22888):   at void org.godotengine.godot.Godot.onCreate(android.os.Bundle) (Godot.java:574)
05-27 13:07:34.470: I/zygote64(22888):   at void android.app.Activity.performCreate(android.os.Bundle, android.os.PersistableBundle) (Activity.java:7009)
05-27 13:07:34.470: I/zygote64(22888):   at void android.app.Activity.performCreate(android.os.Bundle) (Activity.java:7000)
05-27 13:07:34.470: I/zygote64(22888):   at void android.app.Instrumentation.callActivityOnCreate(android.app.Activity, android.os.Bundle) (Instrumentation.java:1214)
05-27 13:07:34.470: I/zygote64(22888):   at android.app.Activity android.app.ActivityThread.performLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent) (ActivityThread.java:2731)
05-27 13:07:34.470: I/zygote64(22888):   at void android.app.ActivityThread.handleLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:2856)
05-27 13:07:34.470: I/zygote64(22888):   at void android.app.ActivityThread.-wrap11(android.app.ActivityThread, android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:-1)
05-27 13:07:34.470: I/zygote64(22888):   at void android.app.ActivityThread$H.handleMessage(android.os.Message) (ActivityThread.java:1589)
05-27 13:07:34.470: I/zygote64(22888):   at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:106)
05-27 13:07:34.470: I/zygote64(22888):   at void android.os.Looper.loop() (Looper.java:164)
05-27 13:07:34.470: I/zygote64(22888):   at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:6494)
05-27 13:07:34.470: I/zygote64(22888):   at java.lang.Object java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[]) (Method.java:-2)
05-27 13:07:34.470: I/zygote64(22888):   at void com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run() (RuntimeInit.java:438)
05-27 13:07:34.470: I/zygote64(22888):   at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:807)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): Unable to load Godot plugin GodotPayment
05-27 13:07:34.475: W/GodotPluginRegistry(22888): java.lang.ClassNotFoundException: org.godotengine.godot.plugin.payment.GodotPayment
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at java.lang.Class.classForName(Native Method)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at java.lang.Class.forName(Class.java:453)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at java.lang.Class.forName(Class.java:378)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at org.godotengine.godot.plugin.GodotPluginRegistry.loadPlugins(GodotPluginRegistry.java:173)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at org.godotengine.godot.plugin.GodotPluginRegistry.<init>(GodotPluginRegistry.java:71)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at org.godotengine.godot.plugin.GodotPluginRegistry.initializePluginRegistry(GodotPluginRegistry.java:103)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at org.godotengine.godot.Godot.onCreate(Godot.java:574)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at android.app.Activity.performCreate(Activity.java:7009)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at android.app.Activity.performCreate(Activity.java:7000)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1214)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2731)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at android.app.ActivityThread.-wrap11(Unknown Source:0)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at android.os.Handler.dispatchMessage(Handler.java:106)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at android.os.Looper.loop(Looper.java:164)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at android.app.ActivityThread.main(ActivityThread.java:6494)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at java.lang.reflect.Method.invoke(Native Method)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): Caused by: java.lang.ClassNotFoundException: org.godotengine.godot.plugin.payment.GodotPayment
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at java.lang.VMClassLoader.findLoadedClass(Native Method)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at java.lang.ClassLoader.findLoadedClass(ClassLoader.java:738)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at java.lang.ClassLoader.loadClass(ClassLoader.java:363)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	... 20 more
05-27 13:07:34.475: W/GodotPluginRegistry(22888): Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Lcom/android/billingclient/api/PurchasesUpdatedListener;
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	... 20 more
05-27 13:07:34.475: W/GodotPluginRegistry(22888): Caused by: java.lang.ClassNotFoundException: Didn't find class "com.android.billingclient.api.PurchasesUpdatedListener" on path: DexPathList[[zip file "/data/app/com.test-s1tfq_9jXDaeuohJv90BvQ==/base.apk"],nativeLibraryDirectories=[/data/app/com.test-s1tfq_9jXDaeuohJv90BvQ==/lib/arm64, /data/app/com.test-s1tfq_9jXDaeuohJv90BvQ==/base.apk!/lib/arm64-v8a, /system/lib64, /vendor/lib64]]
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:125)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
05-27 13:07:34.475: W/GodotPluginRegistry(22888): 	... 20 more
```

**Comment by timoschwarzer** [comment]
@volzhs have you recompiled your editor? You need to do this because the dependencies for the included Android plugins changed.

**Comment by volzhs** [comment]
@timoschwarzer yes. I've always compiled both editor and template.

**Comment by timoschwarzer** [comment]
@volzhs just to be sure: have you re-installed your Android build template?

**Comment by volzhs** [comment]
@timoschwarzer yeap. replaced new templates every time on custom directory path.

**Comment by timoschwarzer** [comment]
I just cloned the repository into a new directory (godot3). These are the steps I did and it worked just fine.

```
git clone https://github.com/godotengine/godot godot3
cd godot3
git checkout 757d8b56721f856e4bd5ec7f27b8bc619ddc2bca
scons platform=x11 bits=64 tools=yes target=release_debug -j 16
scons platform=android bits=64 target=release_debug android_arch=arm64v8 -j 16
scons platform=android bits=64 target=release android_arch=arm64v8 -j 16
scons platform=android bits=64 target=release_debug android_arch=armv7 -j 16
scons platform=android bits=64 target=release android_arch=armv7 -j 16
cd platform/android/java
./gradlew generateGodotTemplates
cd ../../../

# Copy all ZIP, APK and AAR files from ./bin to ~/.local/share/godot/templates/3.2.2.beta

./bin/godot.x11.opt.tools.64
```

Then I created a new project, clicked Project → Install Android Build Template and added a new main scene with the following script:

```gdscript
extends Node

func _ready():
	print(Engine.has_singleton('GodotPayment'))
```

After that I added an Android export template and checked 'Use custom build' and 'Plugins / Godot Payment'.

image

Finally I checked 'Debug → Deploy with Remote Debug' clicked the Android run button in the top right corner of the editor.

Result:
image

**Comment by timoschwarzer** [comment]
For some reason, your Gradle build does not include the Google Play Billing library, which leads to the error you're getting.

**Comment by volzhs** [comment]
@timoschwarzer should I check `Use custom build` to use Godot payment plugin?

**Comment by timoschwarzer** [comment]
@volzhs You have to because Gradle won't fire up and install the dependencies. But yeah, this should be noted in the docs. I didn't have time to update the docs yet but I'll do that the next days. Thank you for reminding me to add a bold notice for that requirement!

**Comment by akien-mga** [comment]
I missed that initially, do we want to force uses to use a custom build to use `GodotPayment`? If so, it should not be exposed as a togglable plugin when "Use Custom Build" is off.

The alternative is to add the billing library in the base APK I guess? But that would force the `BILLING` permission on?

**Comment by m4gr3d** [comment]
@akien-mga @timoschwarzer Unfortunately yes. It's either making the default Android templates include the `BILLING` permission in addition of the billing library, or requiring the use of custom build. 

It's hard trade off because `Use custom build` is harder than the alternative, but we could argue that is the user is setting up billing, which does require configuration, than configuration the Android build tools should be a small ask.

If we end up requiring `Use Custom Build` for `GodotPayment`, we can decouple it completely (as now the code is still included in the default template) as a standalone plugin.

Do we have a hard date for the deprecation of the previous API, as it would simplify the decision?

**Comment by akien-mga** [comment]
I assume that users who use IAP will often want to use other Android plugins like ad providers or Google Play Services, so it's probably OK to make it conditional to doing a custom build.

But indeed, if this is the new requirement, then decoupling it completely from the one-click deploy APK would be good.

**Comment by timoschwarzer** [comment]
> If so, it should not be exposed as a togglable plugin when "Use Custom Build" is off.

> It's hard trade off because Use custom build is harder than the alternative, but we could argue that is the user is setting up billing, which does require configuration, than configuring the Android build tools should be a small ask.

> If we end up requiring Use Custom Build for GodotPayment, we can decouple it completely

I agree to both of you. My proposal would be adding a field like `requires_custom_build` to `PluginConfig` in `https://github.com/godotengine/godot/blob/3.2/platform/android/plugin/godot_plugin_config.h` and disabling/hiding the checkbox in the project export dialog accordingly.

> If we end up requiring Use Custom Build for GodotPayment, we can decouple it completely (as now the code is still included in the default template) as a standalone plugin.

If we decouple the plugin completely, maybe it's worth setting up a first-party Android plugins repository under the `godotengine` namespace with the same branching scheme as this repo. (master, 3.2 etc.)

**Comment by volzhs** [comment]
I can connect store with GodotPayment by following all @timoschwarzer steps. thanks.
now I need to test rest of functions. like `payment.querySkuDetails` :)

**Comment by m4gr3d** [comment]
> I agree to both of you. My proposal would be adding a field like `requires_custom_build` to `PluginConfig` in `https://github.com/godotengine/godot/blob/3.2/platform/android/plugin/godot_plugin_config.h`, disabling/hiding the checkbox in the project export dialog accordingly.


@timoschwarzer All plugins asides from the built-ins (`GodotPayment` being the only one) require  that `Use custom build` be enabled, so that's already implicit.

Disabling/hiding the checkbox in the project export is not possible with the current implementation. I mulled adding the implementation for it, but decided to delay it since we were already in the `beta` phase, and that's kind of semi-large change if done right.

What we can do instead for now, especially if we go ahead with decoupling the `GodotPayment` plugin, is to add a warning at export time if a plugin is enabled but `Use Custom Build` is disabled.
See [here](https://github.com/godotengine/godot/blob/9aeb9ea826cb7ae939fe25366368e42fc85b52fa/platform/android/export/export.cpp#L1891) for an example by @akien-mga

**Comment by HEAVYPOLY** [comment]
I've exported with use custom builds and godotpayment plugin.  Engine.has_singleton("GodotPayment") returns true,
but it gets stuck if I add setpurchasecallbackid.
This was working fine with GodoyPaymentsV3

Is there a guide somewhere on how to get this working with the new system?

```
if Engine.has_singleton("GodotPayment"):
    payment = Engine.get_singleton("GodotPayment")
    payment.setPurchaseCallbackId(get_instance_id())
```

**Comment by timoschwarzer** [comment]
@HEAVYPOLY not yet. I'll begin updating the example project today. Please take a look at "Example code using the new API" above or ping me on Discord/IRC if you need help :)

**Comment by timoschwarzer** [comment]
@akien-mga Where would be the best place for a migration guide for existing GodotPayment**s** users?

**Comment by akien-mga** [comment]
Probably on the official docs. I think there's a page covering IAP, not sure how much details it has regarding the GodotPayments API but I guess it needs both an update + a section with a migration guide.

**Comment by m4gr3d** [comment]
@timoschwarzer Since 3.2.2.stable is [fast approaching](https://godotengine.org/article/release-candidate-godot-3-2-2-rc-2), is this update to the `GodotPayment` plugin feature complete and stable enough for release with 3.2.2.stable?

**Comment by m4gr3d** [comment]
@timoschwarzer cool!

I've pinged @akien-mga to discuss if we should move the plugin into its own repo prior to releasing `3.2.2.stable`. 
That would allow us to quickly address any issues encountered by users, as well as to quickly perform updates (e.g: [Google Play Billing v3](https://android-developers.googleblog.com/2020/06/meet-google-play-billing-library.html)).

**Comment by timoschwarzer** [comment]
@m4gr3d makes sense IMO since it requires Custom Build anyways.

**Comment by akien-mga** [comment]
@m4gr3d @timoschwarzer Moving it to a separate repo makes sense, since we're breaking compatibility anyway we might as well go all the way. We should do this in coming days, I don't want to delay the 3.2.2-stable release too long. But we can take the time needed to do it right of course.

**Comment by m4gr3d** [comment]
@akien-mga I expect the code migration process to be quick. I can perform the migration, we just need to agree on the new repo name and structure. @timoschwarzer Your input would also be appreciated :)

@timoschwarzer Once the code is migrated, you'll need to update the documentation accordingly (both godot docs, and the new repo's README).

**Comment by akien-mga** [comment]
Naming wise, I don't have a strong opinion, beyond using kebab-case, so maybe something like `godotpayment-android-plugin`?

Note that given the scope of the changes, if you think a different name would be better than GodotPayment, that's also an option.

**Comment by timoschwarzer** [comment]
@akien-mga @m4gr3d what about `godot-google-billing` as repo name and renaming the singleton to `GoogleBilling`? In my opinion `GodotPayment` doesn't make clear that it is Android only at first sight. And the new `GodotPayment` (this PR) it's more or less a plain wrapper around the Google Billing library.

**Comment by m4gr3d** [comment]
@timoschwarzer sounds fine by me. It would have to be `godot-google-play-billing` and `GooglePlayBilling` though since this is how it's referred to in the documentation.

**Comment by timoschwarzer** [comment]
@m4gr3d yeah, Google Play Billing of course. (it's late here :sweat_smile:)

**Comment by m4gr3d** [comment]
@timoschwarzer I've created the repo and sent you an invite. Feel free to start updating the repo description and README. 
For the repo files updates, use pull requests so changes can be reviewed and we can ensure we're not stepping over each other.

**Comment by akien-mga** [comment]
Sounds good to me.

@m4gr3d Feel free to start the migration then. I think you should be able to create a new repo under the @godotengine org, if not, you can make it in your namespace and we'll see how to move it when I'm back online tomorrow :)

**Comment by timoschwarzer** [comment]
@m4gr3d @akien-mga Thank you! I'll work on that tomorrow. I'll create an additional PR in the main repository that removes the plugin from the tree.

**Pull request merged**